src/Hashids.net/Hashids.cs

```diff
@@ -274,6 +271,103 @@ public virtual string DecodeHex(string hash)
+
+            var lottery = alphabet[(int)(numberHashInt % _alphabet.Length)];
+            result[0] = lottery;
+
+            var shuffleBuffer = _alphabet.Length < 512 ? stackalloc char[_alphabet.Length] : new char[_alphabet.Length];
+            shuffleBuffer[0] = lottery;
+            _salt.AsSpan().Slice(0, Math.Min(_salt.Length, _alphabet.Length - 1)).CopyTo(shuffleBuffer.Slice(1));
+
+            var startIndex = 1 + _salt.Length;
+            var length = _alphabet.Length - startIndex;
+
+            // use buffer size of 19 which is the length of the biggest 64-bit integer (long.MaxValue = 9223372036854775807)
+            Span<char> hashBuffer = stackalloc char[19];
+
+            if (length > 0)
+                alphabet.Slice(0, length).CopyTo(shuffleBuffer.Slice(startIndex));
+
+            ConsistentShuffle(alphabet, shuffleBuffer);
+            var hashLength = BuildReversedHash(number, alphabet, hashBuffer);
+
+            // Reverse hashBuffer in a loop and insert into result
+            for (var i = 0; i < hashLength; i++)
+                result[i + 1] = hashBuffer[hashLength - i - 1];
+
+            hashLength += 1;
+
+            if (hashLength < _minHashLength)
+            {
+                var guardIndex = (numberHashInt + result[0]) % _guards.Length;
+                var guard = _guards[guardIndex];
+
+                result.Slice(0, hashLength).CopyTo(result.Slice(1));
+                result[0] = guard;
+                hashLength += 1;
+
+                if (hashLength < _minHashLength)
+                {
+                    guardIndex = (numberHashInt + result[2]) % _guards.Length;
+                    guard = _guards[guardIndex];
+
+                    result[hashLength] = guard;
+                    hashLength += 1;
+                }
+            }
+
+            var halfLength = _alphabet.Length / 2;
+
+            var stringBuilder = StringBuilderPool.Get();
+#if NETSTANDARD2_0
+            stringBuilder.Append(result.Slice(0, hashLength).ToArray());
+#else
+            stringBuilder.Append(result[..hashLength]);
+#endif
+
+            while (stringBuilder.Length < _minHashLength)
+            {
+                alphabet.CopyTo(shuffleBuffer);
+                ConsistentShuffle(alphabet, shuffleBuffer);
+
+#if NETSTANDARD2_0
+                stringBuilder.Insert(0, alphabet.Slice(halfLength, _alphabet.Length - halfLength).ToArray());
+                stringBuilder.Append(alphabet.Slice(0, halfLength).ToArray());
+#else
+                stringBuilder.Insert(0, alphabet[halfLength.._alphabet.Length]);
+                stringBuilder.Append(alphabet[..halfLength]);
+#endif
+
+                var excess = stringBuilder.Length - _minHashLength;
+                if (excess > 0)
+                {
+                    stringBuilder.Remove(0, excess / 2);
+                    stringBuilder.Remove(_minHashLength, stringBuilder.Length - _minHashLength);
+                }
+            }
```

Comment:
I was unsure how to implement the remove excess that basically seems to reverse the alphabet but again unsure how to appropriately do that with pure spans and limited space for the result.

The ValueStringBuilder would be great.